### PR TITLE
MV tweedie and D-M for sizefreq

### LINKED ISSUE #66: Make Dirichlet-Multinomial likelihood available for generalized size frequency data
Imported from redmine, Issue [#74989](https://vlab.ncep.noaa.gov/redmine/issues/74989)
 Opened by @iantaylor-NOAA on 2020-02-14
Status when imported: New
 
This task was pulled out of issue #28432 as it may take longer than the other pieces of that issue and thus may not get done for 3.30.15

### LINKED ISSUE #185: consider new informative names for DM parameters
See more comments in #183 . Here is one suggestion:

Couldn't you use a delimiter and just paste the fleets together if there are multiple? e.g., ln(DM_theta)_Age_P7(2-3-4)
For what it is worth, I like the correct mathematical notation of ln(DM_theta) because it helps the visually impaired when a screen reader is reading the output. I think there are ways in tidyverse to accommodate the parenthesis, at least you can with backticks but I think there are other ways as well.

_Originally posted by @kellijohnson-NOAA in https://github.com/nmfs-stock-synthesis/stock-synthesis/issues/183#issuecomment-861718915_

### LINKED ISSUE #219: bin-specific comp logL wrong in compreport for D-M; make comp logL code modular with function
Thanks Grant Thompson for reporting the following:

> ...I was hoping that the log likelihoods listed in CompReport.sso for a given year/fleet/kind would sum to the value listed in the FIT_LEN_COMPS or FIT_AGE_COMPS section of Report.sso, but they are not even close.
> 
> I am using the Dirichlet-multinomial option, and I note that the SS change log includes an entry for 4/5/21 referencing "change logL for D-M composition," so I was wondering if perhaps the change did not get carried over into the code that generates the CompReport.sso file (just a wild guess on my part). ...

Investigate this and see if that is what is going on. (KD: we don't test CompReport.sso at all in our testing routines, so this is certainly possible)

### LINKED ISSUE #360: [Bug]: erroneous effective sample size reported with D-M method 2
### Describe the bug

The code for calculation of D-M logL is correct for but method 1 and method 2.

However, for calculation of effective sample size and generating bootstrap data, method 2 is multiplying the estimated parameter by the input sample size to calculate the Dirichlet_parm used in the effective N calculations.

### To Reproduce

Reported effective N for method 2 is nearly same as the input sample size.  It should be much less unless beta is near 1000.  The error multiples beta by the input sample size, so produces an effective beta near 100 and results in incorrect output sample size.
@iantaylor-NOAA @kellijohnson-NOAA   @Cole-Monnahan-NOAA 

### Expected behavior

see above

### Screenshots

_No response_

### Which OS are you seeing the problem on?

_No response_

### Which version of SS3 are you seeing the problem on?

all

### Additional Context

_No response_

### LINKED ISSUE #362: [Bug]:  offset for length and agecomp using Dirichlet was incorrect for obs with males and females
### Describe the bug

            offset_a(f, i) += gammln(nsamp_a(f, i) + 1.) -
                sum(gammln(1. + nsamp_a(f, i) * obs_a(f, i)(z1, z2)));

The first part of this eq was being added once for females and then again for males.
This occurs only with Dirichlet and only when the comp sample had both males and females
It was detected as a problem when comparing the logL to SizeComp method which does females and males in one calculation.

### To Reproduce

per above

### Expected behavior

expected length comp approach and sizefreq approach to be able to give identical result when configured the same

### Screenshots

_No response_

### Which OS are you seeing the problem on?

_No response_

### Which version of SS3 are you seeing the problem on?

3.30.19 and earlier

### Additional Context

_No response_

### LINKED ISSUE #365: `niter` is only available via `write_SS_summary()` and not `write_big_report()`
https://github.com/nmfs-stock-synthesis/stock-synthesis/blob/0adb061cf7d7fb635ea34b346ed53b9894c5f7ad/SS_write.tpl#L388

Any objection to adding the number of iterations to Report.sso?

### PR DESCRIPTION
**Pull request by Rick-Methot-NOAA**
## Concisely (20 words or less) describe the issue
Reorganizes index logic for comp-weighting parameters in anticipation of the 2-parm MV-Tweedie;
adds information to the Fit_Len_Comp, etc. tables in report.sso, so needs r4ss attention
preps for extending D-M and MV-Tweedie to generalized size comp
## Please Link issue(s)
still open  #266 
resolves #365  resolves #185 resolves #66  resolves #360 resolves #362 resolves #219 

## What tests have been done? Upload any model input files created for testing in a zip file, if possible.
using Andre_tag as testbed.  Also all_gen_size_comp modified to have a combo of length, age and sizecomp, and a combo of multinomial and Dirichlet-Multinomial (D-M).
## What tests/review still need to be done? Who can do it, and by when is it needed (ideally)?
r4ss needs updating to read the revised Fit_Len, etc. tables

## Has any new code been documented?
Yes.
If not, please add documentation before submitting the Pull Request.
- [x] I have documented any new code added (or no new code was added)

## is there an input change for users to Stock Synthesis? 

- [x] Yes, there was an input change

If so, please provide an example of the new inputs needed.

```
-1 # N sizefreq methods to read (or -1 for expanded options)
2 # N sizefreq methods to read (this is a conditional line invoked by -1 above)
HASH each row below has entry for each sizefreq method 
 25 25 #Sizefreq N bins
 2 2 #Sizetfreq units(1=bio/2=num)
 3 3 #Sizefreq scale(1=kg/2=lbs/3=cm/4=inches)
 1e-07 1e-07 #Sizefreq:  small constant to add to comps
 31 49 #Sizefreq N obs per method
 1 0 #_Comp_Error:  0=multinomial, 1=dirichlet using Theta*n, 2=dirichlet using beta, 3=MV_Tweedie  (new conditional line)
 1 0 #_Comp_Error: index for dirichlet or MV_Tweedie  (new conditional line )

```

## Check which is true. This PR requires:

- [ ] no further changes to r4ss
- [ ] no further changes to the manual
- [ ] no further changes to SSI (the SS3 GUI)
- [ ] no further changes to the stock synthesis change log (new features, bug reports)

## Describe any changes in r4ss/SS3 manual/SSI that are needed (if not checked):
r4ss needs to update processing of the "fit_comp" tables
## If changes are needed in the change log, please fill in the table here:

| Action                | Topics                                     | Type  |
| --------------------- | ----------------------------------------- | --------------------------------------- |
| revise | #185 parameter names for D-M | output |
| new | #66 D-M logL added for sizefreq data | ALL |
| fix | #360 D-M reported sample size size | output |
| fix | #362  correct the D-M logL calculation when there are two sexes | output |
| fix | #219  compreport.sso now contains bin-specific logL that is consistent with total logL | output |
## Additional information (optional):
Add functions Comp_logL_multinomial and Comp_logL_Dirichlet to SS_miscfxn.tpl.
These are called by SS_objfun to compute lengthcomp, agecomp and sizefreq logL.

Indexing of CompErr is now by number of invoked methods rather than number of parameters because Tweedie takes 2 parameters

Indexing of CompErr for Sizefreq is by SizeFreq method, not by fleet, 
and this indexing is appended to the fleet-based indexing for lencomp and agecomp.

Add code for computing comp logL by Tweedie.  This is CompErr == 3.  Not operational yet.  Need to move into a function like Comp_Err_Dirichlet().

Add Dirichlet capability to sizefreq

In preliminary calcs.  correct the offset_l and offset_a calculations which were adding a constant twice (for females and again for males).

In readcontrol, update parameter indexing for D-M in sizefreq; revise the CompErr parameter names; 
drop "fleet" from CompErr parameter names because more than 1 fleet can use a parameter.

In write report:  major updating to the Fit_Len, Fit_Age, and Fit_Size tables.  r4ss needs to check for new column names.

Compreport.sso updated such that the reported logL for each bin now sum to the total logL for that observation as reported in the Fit_Len, etc. tables.  This is for multinomial and for D-M.

In write_ssnew:  fix a problem with the sample size used for Dirichlet method 2 (Beta).

### CODE REVIEW

**Comment by Rick-Methot-NOAA** [comment]
Output format looks like this:
Length_Comp_Fit_Summary
Factor Fleet Recommend_var_adj # N Npos min_Nsamp max_Nsamp mean_Nsamp_in mean_Nsamp_adj mean_Nsamp_DM err_method err_index DM_theta MV_power par1 par2 mean_effN HarMean_effN Curr_Var_Adj Fleet_name
4 2 1 # 62 62 25 25 50 25 24.8394 1 1 148.411 NA ln(DM_theta_1)_Len_P1(2) NA 159.429 82.8041 0.5 cpue2
4 4 1 # 10 10 25 25 50 25 24.8394 1 1 148.411 NA ln(DM_theta_1)_Len_P1(2) NA 209.852 68.6758 0.5 cpue4

Note that both fleets share the same comp_err method and parameters.
The parameter name is created for the first fleet that uses it, so that fleet gets included in the label (2).
This makes display of that same parameter label misleading when shown for the second fleet that uses that comp_err method.
I propose leaving "fleet" off of the label for the comperr parameters.

**Comment by k-doering-NOAA** [comment]
I checked out why the call-build-ss3-warnings job is failing, and it is just this one:

```
      |                         ^~~~~~~~~
ss.cpp: In member function ‘void model_parameters::evaluate_the_objective_function()’:
ss.cpp:27450:36: warning: implicitly-declared ‘dvariable& dvariable::operator=(const dvariable&)’ is deprecated [-Wdeprecated-copy]
27450 |     tweedie_Phi = mfexp(selparm(k1));
      |                                    ^
```

I think this warning can be safely ignored, so we would just increase the number of warnings expected by 1 (from 83 to 84): https://github.com/nmfs-stock-synthesis/workflows/blob/a62a6c65cd146ceebd3cac4b612cf3d86c0d58d9/.github/workflows/build-ss3-warnings.yml#L64

**Comment by Rick-Methot-NOAA** [comment]
The one run that failed the test was because there is one fewer warning with new code and that warning was related to a poor search convergence.  So new code is fine.

**Comment by iantaylor-NOAA** [comment]
I'll get to work on reviewing this today.
I can push a change to the warnings test to this branch to make the warning go away.

**Comment by Rick-Methot-NOAA** [comment]
still need to add a base offset for use of D-M with sizefreq method; then do a comparison test between length-based DM and sizefreq-based DM.  Also, last push still had some incorrect effN calcs for method 2 D-M

**Comment by iantaylor-NOAA** [comment]
@Rick-Methot-NOAA, I'm finally done with other meetings.  Should I hold off on review until you've corrected Method 2 effN and added offset or just review what's currently in place?

**Comment by Rick-Methot-NOAA** [comment]
The attached example replicates all length comp data as sizefreq data, assigned both to use the same D-M parameter. Result is the same logL for both length comp and sizefreq comp.
[D-M_test.zip](https://github.com/nmfs-stock-synthesis/stock-synthesis/files/9262085/D-M_test.zip)
Still to be done is work on output to the Fit_SizeFreq table.  This may require creation of an offset value for each sizefreq observation, as is done for length comp

File: SS_objfunc.tpl
Comment (iantaylor-NOAA) [inline SS_objfunc.tpl:650]: Would it make sense to stop with an error here, or does Tweedie for generalized size comp get flagged earlier? If you're planning on adding this option before merging, then no need.

**Review by iantaylor-NOAA — DISMISSED** [review]
@Rick-Methot-NOAA, I looked over all the changes and as far as I can tell things look good although I haven't tried to check the math or compared directly to the equations.

Clearly adding a new likelihood isn't the hard part of this, it's all the variable management and the changes to input/output.

The next steps that I can contribute to would be 
- [x] play around with the the D-M_test.zip model that you attached to the Pull Request so I understand things better
- [x] implement the necessary changes in a branch of r4ss that could be used in the testing
- [ ] compare parameter estimates for existing models that use DM likelihood like the two hake models in https://github.com/nmfs-stock-synthesis/test-models/tree/main/models. Having the r4ss working again may allow those comparisons to happen automatically via github actions. If models change because of errors in the previous implementation, we should try to document that, but maybe it's just the likelihood values that were wrong, not the MLE estimates.
- [ ] update the documentation to describe these changes
- [ ] _[lower priority than the others]_ add a model to the test set that uses the Tweedie distribution for age or length and/or DM with generalized size comp

Having said all that, I'm on leave next week, so please proceed as you wish and I'll figure out how to contribute when I get back.

File: SS_objfunc.tpl
Comment (Rick-Methot-NOAA) [reply SS_objfunc.tpl:650]: I was planning to add Tweedie after we update to ADMB 13.0.  For now, no one knows it is even here, so little risk.

File: SS_objfunc.tpl
Comment (Rick-Methot-NOAA) [reply SS_objfunc.tpl:362]: Agree that this is awkward.  I am leaving these as in for now and relying on the indent of the second line to display the intent.  @nschindler-noaa  Neal - please advise regarding this style issue.

Comment (nschindler-noaa) [reply SS_objfunc.tpl:362]: I think that relying on indents is risky especially when tabs are mixed with spaces. (I know we all use spaces, but my text editor sometimes puts in tabs). And I thought we agreed that single line conditionals and loops were 'out'. So, I like using brackets. Intent is clear, not inferred. Most editors will add a closing bracket when the opening one is put in.

Comment (Rick-Methot-NOAA) [reply SS_objfunc.tpl:362]: @nschindler-noaa 
I think many of the indent-based conditionals were created from single-line conditionals  by the initial application of the style tool.
I agree that always using {} is a good practice

**Comment by k-doering-NOAA** [comment]
@Rick-Methot-NOAA look like the for the run with no est, the two_morph_seas_areas model could not run a second time when using the .ss_new files. maybe the .ss_new file is not being written correctly?

The r4ss job is failing because r4ss is not able to read the output correctly. Maybe @kellijohnson-NOAA can help with this?

The two added warnings are:
```
ss.cpp: In member function ‘dvariable model_parameters::Comp_logL_multinomial(const double&, const dvector&, const dvar_vector&)’:
ss.cpp:18242:45: warning: implicitly-declared ‘dvariable& dvariable::operator=(const dvariable&)’ is deprecated [-Wdeprecated-copy]
18242 |     logL = - Nsamp * obs_comp * log(exp_comp);
      |                                             ^

ss.cpp:27551:36: warning: implicitly-declared ‘dvariable& dvariable::operator=(const dvariable&)’ is deprecated [-Wdeprecated-copy]
27551 |     tweedie_Phi = mfexp(selparm(k1));
      |                                    ^
```

These can be safely ignored, I think, as they are coming from admb 13.

Also, doesn't this PR require admb 13? We still aren't using admb 13 on all the github actions yet. I think this is something Ian is planning to update when he gets back.

**Comment by Rick-Methot-NOAA** [comment]
This PR does not need ADMB 13 yet.  All references to Tweedie are placeholders.

I will look again at the two morph run, but I previously determined that the base run was not fully converged and the new run was slightly better.

**Comment by k-doering-NOAA** [comment]
I think the data_echo.ss_new (or one of the other .ss_new files) may not be writing correctly when running with no estimation.

See this error: https://github.com/nmfs-stock-synthesis/stock-synthesis/runs/7700450307?check_suite_focus=true#step:12:1112

**Comment by Rick-Methot-NOAA** [comment]
found glitch in ss_new code.  fixing now

**Comment by Rick-Methot-NOAA** [comment]
Input format for use of D-M with general sizefreq:
# Sizefreq data. Defined by method because a fleet can use multiple methods
-1 # N sizefreq methods to read (or -1 for expanded options)
2 # N sizefreq methods to read
# each row below has entry for each sizefreq method 
 25 25 #Sizefreq N bins
 2 2 #Sizetfreq units(1=bio/2=num)
 3 3 #Sizefreq scale(1=kg/2=lbs/3=cm/4=inches)
 1e-07 1e-07 #Sizefreq:  small constant to add to comps
 31 49 #Sizefreq N obs per method
 1 0 #_Comp_Error:  0=multinomial, 1=dirichlet using Theta*n, 2=dirichlet using beta, 3=MV_Tweedie
 1 0 #_Comp_Error: index for dirichlet or MV_Tweedie

**Comment by Rick-Methot-NOAA** [comment]
finish refactoring code for compreport.sso to complete the remaining issue #219 
lencomp, agecomp and szcomp now use streamlined and nearly identical code sequence.
previously, the logL values produced for D-M samples were incorrectly based on multinomial, but probably were not being used.
with this code conversion, a test case confirms that the sum of the logL for each bin in compreport.sso matches the total logL for that observation as reported in the report.sso tables Fit_Len, Fit_Age, and Fit_Size. Alas, I did not get this finished before Grant retired.

File: SS_write_ssnew.tpl
Comment (iantaylor-NOAA) [inline SS_write_ssnew.tpl:1272]: Can we change the comment from
"#_Comp_Error: index for dirichlet or MV_Tweedie"
to 
"#_ParmSelect:  consecutive index for dirichlet or MV_Tweedie"
to avoid confusion among these two new inputs lines and better match the comments for other data types?

**Comment by Rick-Methot-NOAA** [comment]
for sure.  glad to have your sharp eyes looking this over.

File: SS_write_ssnew.tpl
Comment (Rick-Methot-NOAA) [reply SS_write_ssnew.tpl]: done, plus some additional related changes to those sections

**Comment by iantaylor-NOAA** [comment]
Changes to r4ss now allow reading the SS3 output associated with this branch (confirmed by the "test-r4ss-with-ss3" workflow with help from @k-doering-NOAA). 
The github actions that run Stock Synthesis with and without estimation needed a little tuning up to compare values from models before and after the changes in this Pull Request but @k-doering-NOAA have now sorted that out and the tests are passing when triggered manually for this branch. I'm going to make a trivial commit to this branch which will trigger a re-run of all the tests within this Pull Request to confirm that result. I think the warnings test will still fail but the others should now pass.

**Comment by iantaylor-NOAA** [comment]
Following up on the fixed tests: now only the warnings test is failing (as expected as discussed in #369), so next steps can be to focus on documentation (maybe in combination with https://github.com/nmfs-stock-synthesis/doc/issues/123) and any needed changes to the SSI and change log.

**Comment by Rick-Methot-NOAA** [comment]
I think if you approve and merge, the change log will automatically get updated

**Review by iantaylor-NOAA — APPROVED** [review]
@Rick-Methot-NOAA, it seems that the additional commits that occurred after my previous review made the approval no longer apply. The additional changes look good, so this approval should allow you to merge whenever you're ready and take away the "Merging is blocked" message.

**Comment by Rick-Methot-NOAA** [comment]
This code revision introduced errors to *.ssnew for sizecomp data.  These are now being corrected with issue #541

**Pull request merged**
